apps/sgx/enclave.cc

```diff
@@ -1,62 +1,19 @@
 #include <dlpack/dlpack.h>
+#include <tvm/runtime/sgx/trusted.h>
+#include <tvm/runtime/c_runtime_api.h>
 #include <tvm/runtime/module.h>
 #include <tvm/runtime/registry.h>
 #include <tvm/runtime/packed_func.h>
-#ifndef _LIBCPP_SGX_CONFIG
-#include <iostream>
-#endif
-
-/* This function mirrors the one in howto_deploy except without the iostream */
-int Verify(tvm::runtime::Module mod, std::string fname) {
-  // Get the function from the module.
-  tvm::runtime::PackedFunc f = mod.GetFunction(fname);
-
-  // Allocate the DLPack data structures.
-  DLTensor* x;
-  DLTensor* y;
-  int ndim = 1;
-  int dtype_code = kDLFloat;
-  int dtype_bits = 32;
-  int dtype_lanes = 1;
-  int device_type = kDLCPU;
-  int device_id = 0;
-  int64_t shape[1] = {10};
-  TVMArrayAlloc(shape, ndim, dtype_code, dtype_bits, dtype_lanes,
-                device_type, device_id, &x);
-  TVMArrayAlloc(shape, ndim, dtype_code, dtype_bits, dtype_lanes,
-                device_type, device_id, &y);
-  for (int i = 0; i < shape[0]; ++i) {
-    static_cast<float*>(x->data)[i] = i;
-  }
-
-  // Invoke the function
-  f(x, y);
-
-  // check the output
-  bool all_eq = true;
-  for (int i = 0; i < shape[0]; ++i) {
-    all_eq = all_eq && static_cast<float*>(y->data)[i] == i + 1.0f;
-  }
-
-  return all_eq;
-}
+#include "../../src/runtime/sgx/runtime_t.cc"
 
+using namespace tvm::runtime;
 
 extern "C" {
-void tvm_ecall_run_module(const void* tvm_args, void* tvm_return_value) {
-  tvm::runtime::Module mod_syslib = (*tvm::runtime::Registry::Get("module._GetSystemLib"))();
-  *(int*)tvm_return_value = Verify(mod_syslib, "addonesys");
+void ecall_tvm_main(const void* args, const int* type_codes, int num_args) {
+  Module mod = (*Registry::Get("module._GetSystemLib"))();
+  PackedFunc f = mod.GetFunction("addonesys");
```

Comment:
In the ideal case, maybe we don't need the enclave.cc at all, and just allow exposure of certain function handles to the outside(possibly by encryption, or have a SGX enclave thread local function table, which you do dispatch)

Reply:
to do so, we need 
```c++
// get function index given name
// return function index, backed by a thread local function table.
int ecall_get_function(const char*name)
// call the function
void ecall_call_function(int func_id, const void* args, const int* type_codes, int num_args);
```

Reply:
We can also support a simple return value system. By supporting a function like ocall_func_set_return. By default if function returns None, then no call is needed.

Reply:
Going the ecall-proxy route would be fine for vanilla TVM modules but not NNVM graphs. An ecall takes ~100x as long as a standard function call (around 1m cycles), so performance requires that the graph runtime be contained in-enclave (the [`Run`](https://github.com/dmlc/tvm/blob/master/src/runtime/graph/graph_runtime.cc#L55) method, specifically).

The good news is that `enclave.cc` can be generated similarly to the cuda host harness. This could be all be done via a python interface. I'd like to do this after getting an enclave model to train using MXNet.

Reply:
That is what I mean, by constraint the graph runtime in-enclave, we can load the enclave module(which contains the graph) and do module.get_function("run") etc. and call its content

Reply:
That is, to support an enclave that allows us to query arbitrary registered TVM function inside that can be exposed, instead of only going through the ecall_main